### PR TITLE
[GlobalISel] Add G_PREFETCH

### PR DESCRIPTION
**Pull request by jayfoad**

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp [hunk 1]
```diff
@@ -2435,6 +2435,18 @@ bool IRTranslator::translateKnownIntrinsic(const CallInst &CI, Intrinsic::ID ID,
...
   }
+  case Intrinsic::prefetch: {
+    Value *Addr = CI.getOperand(0);
+    ConstantInt *RW = cast<ConstantInt>(CI.getOperand(1));
+    ConstantInt *Locality = cast<ConstantInt>(CI.getOperand(2));
+    ConstantInt *CacheType = cast<ConstantInt>(CI.getOperand(3));
+
+    MIRBuilder.buildPrefetch(getOrCreateVReg(*Addr), RW->getZExtValue(),
+                             Locality->getZExtValue(),
+                             CacheType->getZExtValue());
```
Comment (jayfoad) [inline llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:2446]: Should this also get a `MachineMemOperand`?

File: llvm/include/llvm/Target/GenericOpcodes.td [hunk 2]
```diff
@@ -1209,6 +1209,15 @@ def G_FENCE : GenericInstruction {
   let hasSideEffects = true;
 }
 
+// Generic opcode equivalent to the llvm.prefetch intrinsic.
+def G_PREFETCH : GenericInstruction {
+  let OutOperandList = (outs);
+  let InOperandList = (ins ptype0:$address, i32imm:$rw, i32imm:$locality, i32imm:$cachetype);
+  let hasSideEffects = true;
+  let mayLoad = true;
+  let mayStore = true;
```
Comment (jayfoad) [inline llvm/include/llvm/Target/GenericOpcodes.td:1218]: I am not sure about these properties. The DAG opcode has `SDNPHasChain, SDNPMayLoad, SDNPMayStore`. The intrinsic has `DefaultAttrs` and `IntrInaccessibleMemOrArgMemOnly` and `ReadOnly<ArgIndex<0>>`.

Comment (rampitec) [reply llvm/include/llvm/Target/GenericOpcodes.td:1218]: I do not really think mayStore is applicable, but for the compatibility with SDag it needs to match. hasSideEffects is probably OK just to reduce possibilities of reordering, but deserves a comment about this, similar to the description of the IntrInaccessibleMemOrArgMemOnly on the intrinsic.

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp (hunk 1 above)
Comment (rampitec) [reply llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:2446]: SDNode has SDNPMemOperand. SDag lowering also creates memory operand, like this:
```
  S_PREFETCH_DATA killed %2:sgpr_64, 0, $sgpr_null, 0 :: (load (s8) from %ir.ptr, addrspace 4)
```

File: llvm/include/llvm/Target/GenericOpcodes.td (hunk 2 above)
Comment (arsenm) [reply llvm/include/llvm/Target/GenericOpcodes.td:1218]: should probably just be hasSideEffects. mayLoad/mayStore imply it needs a memory operand and is an ordered memory reference when it doesn't have one

Comment (arsenm) [reply llvm/include/llvm/Target/GenericOpcodes.td:1218]: I guess it does point to a load address, so no mayStore

Comment (jayfoad) [reply llvm/include/llvm/Target/GenericOpcodes.td:1218]: I have now copied the logic from SDAG that gives it an MMO marked either MOLoad or MOStore depending on the RW flag. I guess that explains why the opcode needs to be both mayLoad and mayStore - unless we split it into separate G_PREFETCH_LOAD/STORE opcodes.

**Comment by davemgreen** [comment]
The AArch64 parts of this look OK to me.

**Review by arsenm — APPROVED** [review]

File: llvm/include/llvm/Target/GenericOpcodes.td (hunk 2 above)
Comment (rampitec) [reply llvm/include/llvm/Target/GenericOpcodes.td:1218]: > should probably just be hasSideEffects. mayLoad/mayStore imply it needs a memory operand and is an ordered memory reference when it doesn't have one

I could argue this is not a memory operation at all as it shall have no visible effects other than access speed, although practically it has ordering. You certainly do not want a prefetch to be moved past the loads which it was supposed to prefetch. I.e. in my view use of both mayLoad and mayStore is justified. Although we need to make sure it is not considered an aliased store or load from the AA point of view.

**Pull request merged**